apm-lambda-extension/logsapi/subscribe.go

```diff
@@ -0,0 +1,111 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package logsapi
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+var listenerHost = "sandbox"
+var logsAPIServer *http.Server
+var logsAPIListener net.Listener
+
+type LogEvent struct {
+	Time      time.Time       `json:"time"`
+	Type      string          `json:"type"`
+	RawRecord json.RawMessage `json:"record"`
+	Record    LogEventRecord
+}
+
+type LogEventRecord struct {
+	RequestId string `json:"requestId"`
+	Status    string `json:"status"`
+}
+
+// Subscribes to the Logs API
+func subscribe(extensionID string, eventTypes []EventType) error {
+	extensions_api_address, ok := os.LookupEnv("AWS_LAMBDA_RUNTIME_API")
+	if !ok {
+		return errors.New("AWS_LAMBDA_RUNTIME_API is not set")
+	}
+
+	logsAPIBaseUrl := fmt.Sprintf("http://%s", extensions_api_address)
+	logsAPIClient, err := NewClient(logsAPIBaseUrl)
+	if err != nil {
+		return err
+	}
+
+	_, port, _ := net.SplitHostPort(logsAPIListener.Addr().String())
+	_, err = logsAPIClient.Subscribe(eventTypes, URI("http://"+listenerHost+":"+port), extensionID)
+	return err
+}
+
+// Starts the HTTP server listening for log events and subscribes to the Logs API
+func Subscribe(ctx context.Context, extensionID string, eventTypes []EventType, out chan LogEvent) (err error) {
+	if checkAwsSamLocal() {
+		return errors.New("Detected sam local environment")
+	}
+	err = startHttpServer(out)
+	if err != nil {
+		return
+	}
+
+	err = subscribe(extensionID, eventTypes)
+	if err != nil {
+		return
+	}
+	return nil
+}
+
+func startHttpServer(out chan LogEvent) (err error) {
```

Comment:
it would probably be a good idea to pass the `context.Context` from the `Subscribe` func into this one, and listen for `<-ctx.Done()` so that you can know when to call `Shutdown()` on the `logsAPIServer`

Reply:
My thought was that you could know when you want to exit the program, and indicate that to the server so it can allow any running requests to finish, instead of cutting them off. But maybe that's not a concern, since you'll shutdown after the lambda function has exited?

Reply:
I'll move this into another issue because I'm not sure what the best approach is here and we are waiting on the other changes in this PR.
Issue: https://github.com/elastic/apm-aws-lambda/issues/143